### PR TITLE
bootcamp-simon park

### PR DESCRIPTION
**Pull request by s355park**

### CODE REVIEW

File: modules/bootcamp/detect_landing_pad.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 """
 import pathlib
 
+from PIL import Image
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:8]: Don't import other libraries, the bootcamp is completable without any additional libraries.

File: modules/bootcamp/detect_landing_pad.py [hunk 2]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
-
+    
         # Use the model's predict() method to run inference
         # Parameters of interest:
         # * source
         # * conf
         # * device
         # * verbose
-        predictions = ...
+        predictions = self.__model.predict(image, conf=0.7)
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:90]: Also specify `source` and `device`.

File: modules/bootcamp/detect_landing_pad.py [hunk 3]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
+        prediction = predictions[0]
+        image_annotated = prediction.plot(conf=True)
+        # check = Image.fromarray(check[..., ::-1])
+        # check.show()
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:94]: Remove commented codes.

File: modules/bootcamp/detect_landing_pad.py [hunk 4]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
 
-        # Get the Result object
-        prediction = ...
+        # # Get the Result object
+        # prediction = ...
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:97]: Remove commented codes.

File: modules/bootcamp/detect_landing_pad.py [hunk 5]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
-        predictions = ...
+        predictions = self.__model.predict(image, conf=0.7)
+        prediction = predictions[0]
+        image_annotated = prediction.plot(conf=True)
+        # check = Image.fromarray(check[..., ::-1])
+        # check.show()
 
-        # Get the Result object
-        prediction = ...
+        # # Get the Result object
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:96]: Why is there double #s?

File: modules/bootcamp/detect_landing_pad.py [hunk 6]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
+        # prediction = ...
 
         # Plot the annotated image from the Result object
         # Include the confidence value
-        image_annotated = ...
+        # image_annotated = ...
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:101]: Remove commented codes.

File: modules/bootcamp/detect_landing_pad.py [hunk 7]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
 
         # Get the xyxy boxes list from the Boxes object in the Result object
-        boxes_xyxy = ...
+        boxes_xyxy = prediction.boxes.xyxy
 
         # Detach the xyxy boxes to make a copy,
         # move the copy into CPU space,
         # and convert to a numpy array
-        boxes_cpu = ...
+        boxes_cpu = boxes_xyxy.cpu().numpy()
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:109]: You didn't detach it.

File: modules/bootcamp/detect_landing_pad.py [hunk 8]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
 
         # Loop over the boxes list and create a list of bounding boxes
         bounding_boxes = []
+        for i in range(0, boxes_cpu.shape[0]):
+            bounding_boxes.append(bounding_box.BoundingBox.create(boxes_cpu[i, :])[1])
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:114]: I don't get what you mean by `bounding_box.BoundingBox.create(boxes_cpu[i, :])[1]`

File: modules/bootcamp/detect_landing_pad.py [hunk 9]
```diff
@@ -79,39 +80,47 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
+        return (bounding_boxes, image_annotated)
         # Hint: .shape gets the dimensions of the numpy array
         # for i in range(0, ...):
             # Create BoundingBox object and append to list
             # result, box = ...
 
-        # Remove this when done
-        raise NotImplementedError
+        # # Remove this when done
+        # raise NotImplementedError
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:124]: Remove this instead of commenting.

File: modules/bootcamp/decision_simple_waypoint.py [hunk 10]
```diff
@@ -37,7 +37,9 @@ def __init__(self, waypoint: location.Location, acceptance_radius: float):
         # ↓ BOOTCAMPERS MODIFY BELOW THIS COMMENT ↓
         # ============
 
-        # Add your own
+        self.has_moved = False
+
+        self.counter = 0
```
Comment (Trotyl15) [inline modules/bootcamp/decision_simple_waypoint.py:42]: Is counter ever used?

File: modules/bootcamp/decision_simple_waypoint.py [hunk 11]
```diff
@@ -67,14 +69,14 @@ def run(self,
...
         # ↓ BOOTCAMPERS MODIFY BELOW THIS COMMENT ↓
         # ============
-
-        # Do something based on the report and the state of this class...
-
-        # Remove this when done
-        raise NotImplementedError
-
+        if report.status == drone_status.DroneStatus.HALTED and not self.has_moved:
+            command = commands.Command.create_set_relative_destination_command(float(self.waypoint.location_x), float(self.waypoint.location_y))
```
Comment (Trotyl15) [inline modules/bootcamp/decision_simple_waypoint.py:73]: Why do you need to class them to float when they are already floats.

File: modules/bootcamp/decision_simple_waypoint.py [hunk 12]
```diff
@@ -67,14 +69,14 @@ def run(self,
...
+            self.has_moved = True
+        elif report.status == drone_status.DroneStatus.HALTED and self.has_moved:
+            command = commands.Command.create_land_command()
```
Comment (Trotyl15) [inline modules/bootcamp/decision_simple_waypoint.py:76]: What if the drone halts for some unrelated reason while on the way to the waypoint? (You do not need to change anything here, just something to think about).

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 13]
```diff
@@ -12,6 +12,7 @@
 from .. import drone_status
 from .. import location
 from ..private.decision import base_decision
+import math
```
Comment (Trotyl15) [inline modules/bootcamp/decision_waypoint_landing_pads.py:15]: Don't import other libraries, the bootcamp is completable without any additional libraries.

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 14]
```diff
@@ -38,11 +39,17 @@ def __init__(self, waypoint: location.Location, acceptance_radius: float):
...
+        self.state = "initiated"
 
         # ============
         # ↑ BOOTCAMPERS MODIFY ABOVE THIS COMMENT ↑
         # ============
 
+    def distance(self,
+                 location1: location.Location,
+                 location2: location.Location) -> float:
+        return (location2.location_x-location1.location_x)**2+(location2.location_y-location1.location_y)**2
```
Comment (Trotyl15) [inline modules/bootcamp/decision_waypoint_landing_pads.py:51]: Good avoiding square roots! (You don't need to change anything here).

Comment (Trotyl15) [inline modules/bootcamp/decision_waypoint_landing_pads.py:51]: Follow this document for styling guidelines. https://uwarg-docs.atlassian.net/wiki/spaces/CV/pages/2253226033/Python+Style+Convention#Spacing

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 15]
```diff
@@ -69,9 +76,26 @@ def run(self,
         # ============
 
         # Do something based on the report and the state of this class...
+        if report.position.location_x == 0 and report.position.location_y == 0 and report.status == drone_status.DroneStatus.HALTED:
+            command = commands.Command.create_set_relative_destination_command(float(self.waypoint.location_x), float(self.waypoint.location_y))
+        elif report.status == drone_status.DroneStatus.HALTED and report.position == self.waypoint:
+            print("reached waypoint")
+            closest_pad_distance = 300
```
Comment (Trotyl15) [inline modules/bootcamp/decision_waypoint_landing_pads.py:83]: What is the clostest landing pad is further than 300

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 16]
```diff
@@ -69,9 +76,26 @@ def run(self,
...
+            closest_pad_index = -1
+            for i in range(len(landing_pad_locations)):
+                distance = self.distance(self.waypoint, landing_pad_locations[i])
+                if distance < closest_pad_distance:
+                    closest_pad_distance = distance
+                    closest_pad_index = i
+            print(f"closest landing pad: {landing_pad_locations[closest_pad_index]}")
```
Comment (Trotyl15) [inline modules/bootcamp/decision_waypoint_landing_pads.py:90]: Good use of commenting! (You don't need to change anything here)

File: modules/bootcamp/decision_waypoint_landing_pads.py [hunk 17]
```diff
@@ -69,9 +76,26 @@ def run(self,
...
+            command = commands.Command.create_set_relative_destination_command((landing_pad_locations[i].location_x-report.position.location_x), (landing_pad_locations[i].location_y - report.position.location_y))
+            self.state = "past waypoint"
+        elif report.status == drone_status.DroneStatus.HALTED and self.state == "past waypoint":
+            command = commands.Command.create_land_command()
+                    
 
         # Remove this when done
-        raise NotImplementedError
+        # raise NotImplementedError
```
Comment (Trotyl15) [inline modules/bootcamp/decision_waypoint_landing_pads.py:98]: Remove instead of commenting.

File: modules/bootcamp/detect_landing_pad.py [hunk 18]
```diff
@@ -99,7 +100,7 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
         # Detach the xyxy boxes to make a copy,
         # move the copy into CPU space,
         # and convert to a numpy array
-        boxes_cpu = boxes_xyxy.cpu().numpy()
+        boxes_cpu = torch.detach(boxes_xyxy).cpu().numpy()
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:103]: `boxes_xyxy.detach().cpu().numpy()`

File: modules/bootcamp/detect_landing_pad.py [hunk 19]
```diff
@@ -79,40 +78,35 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
-        # for i in range(0, ...):
-            # Create BoundingBox object and append to list
-            # result, box = ...
-
-        # Remove this when done
-        raise NotImplementedError
+        for i in range(0, boxes_cpu.shape[0]):
+            bounding_boxes.append(bounding_box.BoundingBox.create(boxes_cpu[i])[1])
 
+        return (bounding_boxes, image_annotated)
```
Comment (Trotyl15) [inline modules/bootcamp/detect_landing_pad.py:109]: In statistics, an invalid portion of the data point involves removing the entirety of the data point (where the data point here is an image, as each image is independent). If the result `(bounding_box.BoundingBox.create(boxes_cpu[i])[0]`) is invalid, return an empty list with `image_annotated`.

File: modules/bootcamp/detect_landing_pad.py [hunk 20]
```diff
@@ -79,40 +78,37 @@ def run(self, image: np.ndarray) -> "tuple[list[bounding_box.BoundingBox], np.nd
...
-            # Create BoundingBox object and append to list
-            # result, box = ...
-
-        # Remove this when done
-        raise NotImplementedError
-
+        for i in range(0, boxes_cpu.shape[0]):
+            result, new_bounding_box = bounding_box.BoundingBox.create(boxes_cpu[i])
+            if not result:
+                return ([], [])
```
Comment (AaronWang04) [inline modules/bootcamp/detect_landing_pad.py:109]: None would be a better return instead of empty lists. it makes it clear that there was an invalid data

**Review by AaronWang04 — COMMENTED** [review]
everything looks good, just a nitpick comment (you don't need to change and recommit, but be aware of it). approved
